util/conditions/patch.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conditions
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+)
+
+// PatchOperation defines patch operation types.
+type PatchOperation string
+
+const (
+	// AddConditionPatch defines an add condition patch operation.
+	AddConditionPatch PatchOperation = "Add"
+
+	// ChangeConditionPatch defines an change condition patch operation.
+	ChangeConditionPatch PatchOperation = "Change"
+
+	// RemoveConditionPatch defines a remove condition patch operation.
+	RemoveConditionPatch PatchOperation = "Remove"
```

Comment:
I think that this util should be generic and it must handle all the possible patch operations because in the future it can be used also outside the Patch helper.

BTW, I remember in the past discussing some use cases where deleting a condition can happen, like e.g. 
KCP upgrade completes and among others it sets CoreDnsUpgradeSucceded,
After a while a new upgrade is restarted; in that case we can potentially be required to remove the old UpgradeSucceded (we are skipping core-DNS upgrade).